### PR TITLE
[spec] Add fixtures for people, emails, and versions.

### PR DESCRIPTION
**Pull request by alloy**
As challenged, here are some fixtures (based on Breaking Bad characters). Will comment inline.

### CODE REVIEW

File: spec/fixtures/emails.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+skyler:
+  person: $LABEL
+  address: A1A-carwash@example.com
+  normalized_address: <%= CanonicalEmail.normalize_address("A1A-carwash@example.com") %>
+
+jesse:
+  person: $LABEL
+  address: Cap'n-Cook@example.com
+  normalized_address: <%= CanonicalEmail.normalize_address("Cap'n-Cook@example.com") %>
```
Comment (alloy) [inline spec/fixtures/emails.yml]: Got an apostrophe in there 😎

File: spec/fixtures/versions.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<% YAML.load(File.read(File.expand_path("../people.yml", __FILE__))).each do |label, hash| %>
```
Comment (alloy) [inline spec/fixtures/versions.yml]: This loops over the people defined in the `people.yml` file and generates a ‘create’ version entry for each.

File: spec/fixtures/versions.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  <% id = ActiveRecord::FixtureSet.identify(label) %>
```
Comment (alloy) [inline spec/fixtures/versions.yml]: Fixtures without an explicit ID get an ID that’s the hash of their label. Seeing as the version explicitly needs the ID in a few places we generate the ID in the same way.

File: spec/models/email_spec.rb [hunk 4]
```diff
@@ -1,27 +1,31 @@
 require 'rails_helper'
 
-RSpec.describe Email, type: :model do
+RSpec.describe Email do
```
Comment (alloy) [inline spec/models/email_spec.rb]: As [pointed out in the configuration](https://github.com/artsy/constellation/blob/3a8474a9c950939a2535f0df0fb2c130716b7bc4/spec/rails_helper.rb#L38-L51) specifying the type is not needed. I like the noise it removes.

File: spec/models/email_spec.rb [hunk 5]
```diff
@@ -1,27 +1,31 @@
...
   context 'validations' do
     it 'is invalid' do
       expect(Email.new(address: 'something@ blah.com')).to_not be_valid
     end
+
```
Comment (alloy) [inline spec/models/email_spec.rb]: I think we talked about this on and off previously, I really prefer newlines in between blocks for readability 😎

File: spec/models/person_spec.rb [hunk 6]
```diff
@@ -1,46 +1,31 @@
...
+        person.update_attributes!(name: 'Percy Zikherman')
       end
-    end
-    context 'when the email does not exist' do
-      it 'creates the email, person and rsvp interaction' do
-        # Create person w/ primary email + name
-        p = Person.create!(name: 'Percy Z')
-        p.emails.create!(address: 'percy@email.com')
-        # Create RSVP interaction
-        p.interactions.create!(interaction_type: 'rsvp', description: 'RSVP to Catty Art Show')
```
Comment (alloy) [inline spec/models/person_spec.rb]: This wasn’t testing anything, so I removed it.

File: spec/models/person_spec.rb [hunk 7]
```diff
@@ -1,46 +1,31 @@
...
-        p = Email.where(address: 'percy@email.com').first.person
-        # Create RSVP interaction
-        i = p.interactions.create!(interaction_type: 'rsvp', description: 'RSVP to Catty Art Show')
-        # Update full name and tie it to RSVP
-        PaperTrail.whodunnit = i
-        p.update_attributes!(name: 'Percy Zikherman')
-        # Check fields/history
-        actor = p.versions.last.actor
-        expect(actor.description).to eq 'RSVP to Catty Art Show'
-        expect(p.paper_trail.previous_version.name).to eq 'Percy Z'
```
Comment (alloy) [inline spec/models/person_spec.rb]: This was testing some things that are unrelated to the `Person` model, I removed those parts and otherwise split the test up in distinct tests.

Comment (alloy) [inline spec/models/person_spec.rb]: I also removed comments in favour of test descriptions.

File: spec/fixtures/emails.yml (hunk 1 above)
Comment (mzikherman) [inline spec/fixtures/emails.yml]: oh noes!

File: spec/fixtures/versions.yml (hunk 2 above)
Comment (mzikherman) [inline spec/fixtures/versions.yml]: I sort of start to wonder about the kosher-ness of this- creating fixtures manually when they happen as part of the callback lifecycle feels iffy. I get that if you wanted to write specs around the paper trail you'd have to actually fabricate/create objects (and let them go thru that lifecycle).

In this case/app, it seems like that's a pretty crucial thing to have happen, so I wonder the benefit of this fixture.

File: spec/models/email_spec.rb (hunk 5 above)
Comment (mzikherman) [inline spec/models/email_spec.rb]: 👍

File: spec/models/email_spec.rb (hunk 4 above)
Comment (mzikherman) [inline spec/models/email_spec.rb]: Oh cool, I didn't realize that.

File: spec/models/person_spec.rb (hunk 6 above)
Comment (mzikherman) [inline spec/models/person_spec.rb]: oh yea, lol

**Comment by joeyAghion** [comment]
I'm keeping an eye on this out of curiosity... I never got into fixtures much in the past but maybe I just never saw someone use them well. :eyes:

**Comment by mzikherman** [comment]
Thanks for doing this!

As I look at the actual working example, I'm not really compelled to use fixtures, and actually would prefer factories at this point.

- YAML ( 😐 )
- custom code for loading papertrail-ed data outside of normal lifecycle

I also suspect we'll be using actual fabricators mostly since all of our integration-y specs and more complicated setups will necessarily need to.

I just don't fully grok the benefit of using fixtures _in addition_ to the factory stuff, esp if it's just some of the really simple unit specs that can use them (which I think is likely). It also seems odd to me (when imagining testing some more complicated scenarios) that there'll be data present that wasn't explicitly setup for that test?

I don't know, I'm not really feeling it...

File: spec/fixtures/emails.yml (hunk 1 above)
Comment (alloy) [inline spec/fixtures/emails.yml]: No this is good, I mean :)

File: spec/fixtures/versions.yml (hunk 2 above)
Comment (alloy) [inline spec/fixtures/versions.yml]: I don’t think we’d be adding any more than just these initial ‘create’ versions. In further testing of the app you would probably create versions by using the normal API.

That fixtures otherwise bypass callbacks is actually in a lot of cases a benefit.

Comment (alloy) [inline spec/fixtures/versions.yml]: And this model is an exceptional case, but you won’t have to think about it anymore because it will just pickup changes to the list of people.

File: spec/fixtures/emails.yml (hunk 1 above)
Comment (joeyAghion) [inline spec/fixtures/emails.yml]: I feel strange about calling the app code within fixtures... shouldn't these just contain the static, normalized values instead?

File: spec/fixtures/versions.yml (hunk 2 above)
Comment (joeyAghion) [inline spec/fixtures/versions.yml]: I might just be spoiled by factories, but I tend to agree about the iffy-ness here. If any spec actually relied on this data or made assertions about them (or any secondary effects of their having been created), it would not be confirming production-like behavior but instead the behavior of this test code. (In a simple case, if we upgraded `paper_trail` or even dropped it altogether, this data would be misleading.)

**Comment by alloy** [comment]
> custom code for loading papertrail-ed data outside of normal lifecycle

This is an exceptional case.

> I also suspect we'll be using actual fabricators mostly since all of our integration-y specs and more complicated setups will necessarily need to.

As discussed, there’s nothing stopping you from creating new specific records where needed.

> It also seems odd to me (when imagining testing some more complicated scenarios) that there'll be data present that wasn't explicitly setup for that test?

That’s actually a more realistic state of the database and you’ll get that populated database without having to do any elaborate setup. Your tests also generally shouldn’t assert absolute things (like `expect(Model.count).to eq(2)`) but rather be relative (so `expect(lambda { … }).to change(Model, :count).by(2)`.

And, sure, in some cases you really do want an empty table, then you’ll use the occasional `Model.delete_all`. It’s definitely not the nicest thing, but all things come with tradeoffs and I think this is a minor one.

> I just don't fully grok the benefit of using fixtures in addition to the factory stuff, esp if it's just some of the really simple unit specs that can use them (which I think is likely).

I disagree that it’s only the really simple unit specs. People have been building large apps for years with fixtures and creating the specific records where needed. E.g. Basecamp.

A test suite should be hella fast so that you enjoy writing many more (unit) tests, especially when you want to do TDD and don’t want to wait too long before writing an implementation. Seeing as factories are an order of magnitude slower,  [as illustrated before](https://gist.github.com/alloy/e67a92f134d6ba7eee2face648a67ec5), using fixtures where you don’t need anything more is a great way to maintain performance without doing much of anything.

Given these two equal examples, where you just need _a_ person, I’d like to flip around your question; why would you _not_ choose for the faster solution?

```ruby
Factory(:person)
```

vs

```ruby
people(:walter)
```

> I don't know, I'm not really feeling it...

That’s why you should give them an honest try and be able to form an opinion based on experience instead ;)

File: spec/fixtures/emails.yml (hunk 1 above)
Comment (alloy) [inline spec/fixtures/emails.yml]: Nope, this is totally encouraged. In fact you can even pull helpers into this context.

**Comment by alloy** [comment]
> maybe I just never saw someone use them well

@joeyAghion It’s definitely something that one must get well at, it has also been a while for me.

File: spec/fixtures/versions.yml (hunk 2 above)
Comment (alloy) [inline spec/fixtures/versions.yml]: If we dropped it then this data would fail to be inserted into the table.

I agree that this is not the nicest thing, but paper_trail and the data it duplicates is an exceptional case.

**Comment by mzikherman** [comment]
> Given these two equal examples, where you just need a person, I’d like to flip around your question; why would you not choose for the faster solution?

@alloy b/c they aren't actually equal examples IMO. One actually creates an object and subjects it to the current code's callback lifecycle, etc. whereas the other one is just reading a YAML file (that may or may not be accurate).

**Comment by alloy** [comment]
Sure, that’s a different case. If you want to test side-effects, then create a record, my point is all about the times where you just need a record to exist. Creating a new record which may trigger side-effects then does more than you’re actually want to be testing.

**Comment by mzikherman** [comment]
> My point is all about the times where you just need a record to exist.

Right, but I think those are only for really trivial unit tests and such, of which we probably won't have many, focusing more on the request/integration/acceptance specs.

And then for unit tests we do have where fixtures might be appropriate since we'd want data to just exist....that just feels like a weird/not common use case to add some additional code and complexity now for? It's just not compelling to me.

> That’s why you should give them an honest try and be able to form an opinion based on experience instead ;)

That's a fair point, hard to argue with that :)

**Comment by bhoggard** [comment]
I love fixtures! It really speeds up complex tests and saves the effort/time of making a lot of objects in the code.

**Comment by alloy** [comment]
Why do you think we would not have many unit tests?

**Comment by mzikherman** [comment]
> Why do you think we would not have many unit tests?

I didn't say that, I said I didn't think would have many _trivial_ unit tests, for where fixtures are appropriate. We'll have tons of unit tests I'm sure, but wouldn't we want those to be subject to the actual lifecycle of objects?

The really trivial ones where we want the data to already exist as fixtures, I don't see tons of those, no.

**Comment by alloy** [comment]
Ah gotcha. Unit tests should test the smallest unit of code, not coincidentally test side-effects that are out of scope for the described test.

**Comment by joeyAghion** [comment]
I love the active debate.

Some of the disagreement seems rooted in past experience as well as different understandings of what this suite (and project, in general) will end up looking like. Since Constellation is committed to being boring in most technology choices, maybe this can be 1 small area of experimentation/risk-taking?

I, for one, am open to being convinced of the value of fixtures with a real-life project. And switching away if the experience is negative is awkward but not insurmountable, especially if the team commits to make a deliberate yes/no decision about it in a few sprints time (say, as part of a retrospective).

By the way, re. unit vs. integration tests, I actually see the fixtures a little differently. I see them as providing a stable, fast pre-setup state for all sorts of tests, but especially non-unit ones. In the Gravity analogy, imagine if tests could skip needing to create a `User`, `Partner`, `Artwork`, `Artist`, `ClientApplication`, etc. Zoom zoom!

**Comment by alloy** [comment]
@joeyAghion Yeah great point, it definitely can simplify setup.

I figured I’d focus on the “if all else is equal, then let’s choose performance” part (because I really do care for faster test suites), but in retrospect that may not have been the best choice to focus on. Thanks for rectifying that 👍

**Comment by mzikherman** [comment]
So shall we then merge and we can make sure to keep it in mind/retrospect on it at some point in the future?

**Comment by alloy** [comment]
Totally, I wouldn’t expect anything less 👍! Anything is debatable, also when merged.
